src/algorithms.hpp

```diff
@@ -283,43 +285,101 @@ bool is_unique(const C& c)
     return std::adjacent_find(c.begin(), c.end()) == c.end();
 }
 
+template <typename R1, typename R2>
+class index_into_iterator:
+    public std::iterator<std::forward_iterator_tag, typename R1::value_type>
+{
+public:
+    using value_type = typename R1::value_type;
```

Comment:
If the class is ultimately going to be defined over iterator types rather than range types, then `iterator_traits<super_iterator>::difference_type` will be the way to go. Analogous to indices for e.g. `std::string`, an unsigned type might be preferable.